cql3/expr/expression.cc

```diff
@@ -2461,5 +2461,15 @@ bool contains_multi_column_restriction(const expression& e) {
     });
     return find_res != nullptr;
 }
+
+bool has_only_eq_binops(const expression& e) {
+    const expr::binary_operator* non_eq_binop = find_in_expression<expr::binary_operator>(e,
+        [](const expr::binary_operator& binop) {
+            return binop.op != expr::oper_t::EQ;
+        }
+    );
+
+    return non_eq_binop == nullptr;
+}
```

Comment:
This is not accurate. partition_key_restrictions_is_all_eq() could assume that the expression is a conjunction of binops. But this generic function can't assume anything that, it's not in partition key restrictions context. A new user could call it and be surprised by the result.

We could fix it with a longer name that will explain the prerequisites: conjunection_of_binops_has_only_eq_binops(), or by checking that it's (a conjunction of) binops.

If we had boolean_factors(expression) we could write

```c++
bool has_only_eq_binops(e) {
   return all_of(boolean_factors(e) | transformed({
          [] (expression e) { return expr::is<binary_operator>(e) && expr::as<binary_operator>(e).op == EQ); }
   })
}
```

It's also more accurate when RHS can support a nested binop: `WHERE A = (? > ?)`.

So, the name should not claim to do more than it actually does.

Reply:
It handles `A = (? > ?)`, and it gives `false`, but the result we want is `true` - we could solve for a partition key or clustering key from that.

Checking the partition key restrictions is only interested in the top-level binary operators.

Reply:
I guess we could say the function is accurate, but its user is not accurate and will have to be adjusted once we relax the grammar.